# **MAE598 Homework 1**

### Q1. Solve the following problem using Python SciPy.optimize or JuMP. Please attach your code and results. Specify your initial guesses of the solution. If you change your initial guess, do you find different solutions?


###Minimize:
$(x1-x2)^2+(x2+x3-2)^2+(x4-1)^2+(x5-1)^2$
###Subject to:
$x1+3x2 = 0

x3+x4-2x5 = 0

x2-x5 = 0$
###Boundary Conditions :
-10<= xi <= 10, i = 1,2,3,4,5
###Initial guesses of the solution
[-34,-75,-16,-97,-28],[1,2,3,4,5],[30,11,52,23,64],[0,0,0,0,0],[-90,-89,-88,-87,-86],[-1,-2,-3,-4,-5]

In [4]:

from scipy.optimize import minimize

# Given Function
fun = lambda x: (x[0] - x[1])**2 + (x[1] + x[2]-2)**2 + (x[3]-1)**2 + (x[4]-1)**2

# Given Constraints
const = ({'type': 'eq', 'fun': lambda x:  x[0] + 3 * x[1]},
        {'type': 'eq', 'fun': lambda x:  x[2]+ x[3] - 2 * x[4]},
        {'type': 'eq', 'fun': lambda x:  x[1]-x[4]})

# Given Boundary Conditions
boundary = ((-10,10),(-10,10),(-10,10),(-10,10),(-10,10))

# Iteration 1
result_1 = minimize(fun, (-34,-75,-16,-97,-28), method='SLSQP', bounds=boundary,constraints=const)

# Output1
result_1
print(result_1)

print('--------------------------------------------------------------------------------------------------------------------------')
print('--------------------------------------------------------------------------------------------------------------------------')
# Iteration 2
result_2 = minimize(fun, (1,2,3,4,5), method='SLSQP', bounds=boundary,constraints=const)

# Output2
result_2
print(result_2)

print('--------------------------------------------------------------------------------------------------------------------------')
print('--------------------------------------------------------------------------------------------------------------------------')
# Iteration 3
result_3 = minimize(fun, (30,11,52,23,64), method='SLSQP', bounds=boundary,constraints=const)

# Output3
result_3
print(result_3)

print('--------------------------------------------------------------------------------------------------------------------------')
print('--------------------------------------------------------------------------------------------------------------------------')
# Iteration 4
result_4 = minimize(fun, (0,0,0,0,0), method='SLSQP', bounds=boundary,constraints=const)

# Output4
result_4
print(result_4)

print('--------------------------------------------------------------------------------------------------------------------------')
print('--------------------------------------------------------------------------------------------------------------------------')
# Iteration 5
result_5 = minimize(fun, (-90,-89,-88,-87,-86), method='SLSQP', bounds=boundary,constraints=const)

# Output5
result_5
print(result_5)

print('--------------------------------------------------------------------------------------------------------------------------')
print('--------------------------------------------------------------------------------------------------------------------------')
# Iteration 4
result_6 = minimize(fun, (-1,-2,-3,-4,-5), method='SLSQP', bounds=boundary,constraints=const)

# Output6
result_6
print(result_6)

 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.09302325604787
       x: [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
     nit: 8
     jac: [-2.047e+00 -1.860e-01 -2.233e+00 -2.233e+00 -1.488e+00]
    nfev: 48
    njev: 8
--------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------
 message: Optimization terminated successfully
 success: True
  status: 0
     fun: 4.09302325588931
       x: [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
     nit: 7
     jac: [-2.046e+00 -1.880e-01 -2.234e+00 -2.232e+00 -1.489e+00]
    nfev: 44
    njev: 7
--------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------

###To observe if change in the initial guess, give us different solutions, I have made a table


Initial condition (Iteration) | Function | X values
--- | --- | ---
[-34,-75,-16,-97,-28] | 4.093023258012615 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[1,2,3,4,5] | 4.09302325588931 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[30,11,52,23,64] | 4.093023258012615 |  [-7.675e-01  2.558e-01  6.279e-01 -1.162e-01  2.558e-01]
[0,0,0,0,0] | 4.093023255813954 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[-90,-89,-88,-87,-86] | 4.09302325604787 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]
[-1,-2,-3,-4,-5] | 4.093023255813954 |  [-7.674e-01  2.558e-01  6.279e-01 -1.163e-01  2.558e-01]



From the above comparison, it can be seen that over a large range of inital conditions, there is very minimum change in the solution.


### Q2. Let x and b ∈ Rn be vectors and A ∈ Rn×n be a square matrix. Define
f : Rn → R as f (x) = b^T x + x^T Ax.

###Answer -

(a) What are the gradient and Hessian of f (x) with respect to x?

Gradient g(x)=∂f(x)/∂x = b+(A+A^T)x

Hessian H(x)= ∂g(x)/∂x = (A+A^T)


(b) Derive the first- and second-order Taylor’s approximations of f (x)
at x = 0. Are these approximations exact?


The first-order Taylor's approximation of a function f(x) at a point x = a is given by:
f(x) ≈ f(a) + ∇f(a)^T (x - a)

The second-order Taylor's approximation includes both linear and quadratic terms:
f(x) ≈ f(a) + ∇f(a)^T (x - a) + 0.5 * (x - a)^T H(a) (x - a)


At x = 0:

First-order Taylor's approximation:
f(x) ≈ f(0) + ∇f(0)^T x = b^T x

Second-order Taylor's approximation:
f(x) ≈ f(0) + ∇f(0)^T x + 0.5 * x^T H(0) x = b^T x + 0.5 * x^T (A + A^T) x


Are these approximations exact?

These approximations are not exact. They are valid when x is close to the point of approximation (x = 0 in this case).





### Q3. Let A ∈ Rn×n be a square matrix

### Answer-

a) What are the necessary and sufficient conditions for A to be pos-
itive definite?

All of the eigen values should be positive, then A is positive definite.

(b) What are the necessary and sufficient conditions for A to have full
rank?

The determinant of A is nonzero, then A will have full rank.

(c) If there exists y ∈ Rn and y 6 = 0 such that AT y = 0, then what
are the conditions on b for Ax = b to have a solution for x?

b should be orthogonal to y for Ax=b to have a solution for x.



### Q4. Due to the recent inflation, let’s reconsider the Stigler diet problem proposed by Nobel laureate George Stigler after the second World War: Consider that the grocery store offers N types of food of your interest, and each food contains the same M types of nutrition. Denote aij as the quantity of nutrition type j of food type i for i = 1, 2, ..., N and j = 1, 2, ..., M , ci the unit price of food type i, and bj the necessary quantity of nutrition type j for a month. Formulate an optimization problem to determine the minimum grocery cost to satisfy the nutrition needs.

### Answer-

**Variables:**

$x_{i}$ as the quantity of food type i to be purchased.

**Objective Fuction:**

Minimize the total cost while meeting the nutritional requirements:

**Minimize:**
$\sum_{i=1}^N c_{i} x_{i}$   

**Constraints:**

*Nutritional Requirements Constraint:*
For each nutritional type j, the total amount of nutrition obtained from the selected foods should meet or exceed the required amount $b_{j}$

$\sum_{i=1}^N a_{ij} x_{i} >= b_{j}$, for $j=1,2,...,M$


*Non-Negativity Constraint:*
The quantity of each food type purchased cannot be negative:
$x_{i} >= 0$, for $i = 1,2,...,N$

The complete optimization problem can be summarized as:

**Minimize:** $\sum_{i=1}^N c_{i} x_{i}$  

*Subject to:*

$\sum_{i=1}^N a_{ij} x_{i} >= b_{j}$, for $j=1,2,...,M$

$x_{i} >= 0$, for $i = 1,2,...,N$


In this problem, $N$ represents the number of food types available, $M$ represents the number of nutritional types, $a_{ij}$  represents the quantity of nutritional type $j$ in food type $i$, $c_{i}$ represents the cost of food type $i$, and $ b_{j}$ represents the required quantity of nutritional type $j$ for a month.

